src/jit/gentree.cpp

```diff
@@ -9183,231 +9183,340 @@ GenTreePtr GenTree::GetChild(unsigned childNum)
-                case 1:
-                    return &m_node->AsArrOffs()->gtIndex;
-                case 2:
-                    return &m_node->AsArrOffs()->gtArrObj;
-                default:
-                    return nullptr;
-            }
+            case GT_ARR_OFFSET:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceArrOffset;
+                break;
+
+            case GT_DYN_BLK:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceDynBlk;
+                break;
+
+            case GT_STORE_DYN_BLK:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceStoreDynBlk;
+                break;
+
+            case GT_LEA:
+                if (m_node->AsAddrMode()->gtOp1 == nullptr)
+                {
+                    m_advance = &GenTreeUseEdgeIterator::AdvanceLea;
+                }
+                else
+                {
+                    m_advance = GetAdvanceFnForBinOp(m_node);
+                }
+                break;
+
+            case GT_PHI:
+                m_argList = m_node->AsUnOp()->gtOp1;
+                m_advance = &GenTreeUseEdgeIterator::AdvanceList;
+                break;
 
-        // Call, phi, and SIMD nodes are handled by MoveNext{Call,Phi,SIMD}UseEdge, repsectively.
-        case GT_CALL:
-        case GT_PHI:
 #ifdef FEATURE_SIMD
-        case GT_SIMD:
+            case GT_SIMD:
+                if (m_node->AsSIMD()->gtSIMDIntrinsicID == SIMDIntrinsicInitN)
+                {
+                    m_argList = node->AsSIMD()->gtOp1;
+                    m_advance = &GenTreeUseEdgeIterator::AdvanceList;
+                }
+                else
+                {
+                    m_advance = GetAdvanceFnForBinOp(m_node);
+                }
+                break;
 #endif
-            break;
 
-        case GT_ASG:
+            case GT_FIELD_LIST:
+                m_argList = node;
+                m_advance = &GenTreeUseEdgeIterator::AdvanceList;
+                break;
+
+            case GT_CALL:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceCall;
+                break;
+
+            default:
+                if (m_node->OperIsUnary())
+                {
+                    GenTree** edge = &m_node->AsUnOp()->gtOp1;
+                    if (*edge == nullptr)
+                    {
+                        m_state = -1;
+                    }
+                    else
+                    {
+                        m_edge = edge;
+                        m_advance = &GenTreeUseEdgeIterator::AdvanceUnary;
+                    }
+                    return;
+                }
+                else
+                {
+                    m_advance = GetAdvanceFnForBinOp(m_node);
+                }
+                break;
+        }
+
+        // Advance to the first operand.
+        ++(*this);
+    }
+}
+
+void GenTreeUseEdgeIterator::AdvanceCmpXchg()
+{
+    FinishSimpleAdvance([this]() -> GenTree** {
+        switch (m_state)
```

Comment:
Prebuilding an array of element pointers looks like this:

```diff
--- a/src/jit/gentree.cpp
+++ b/src/jit/gentree.cpp
@@ -9303,33 +9303,26 @@ GenTreeUseEdgeIterator::GenTreeUseEdgeIterator(GenTree* node)
 
 void GenTreeUseEdgeIterator::AdvanceCmpXchg()
 {
+    static GenTree* GenTreeCmpXchg::*OperandTable[] = {
+        &GenTreeCmpXchg::gtOpLocation,
+        &GenTreeCmpXchg::gtOpValue,
+        &GenTreeCmpXchg::gtOpComparand,
+    };
+
     FinishSimpleAdvance([this]() -> GenTree** {
-        switch (m_state)
-        {
-            case 0:
-                return &m_node->AsCmpXchg()->gtOpLocation;
-            case 1:
-                return &m_node->AsCmpXchg()->gtOpValue;
-            case 2:
-                return &m_node->AsCmpXchg()->gtOpComparand;
-            default:
-                return nullptr;
-        }
+        return m_state > 2 ? nullptr : &(m_node->AsCmpXchg()->*OperandTable[m_state]);
     }());
 }
```

Here's a standalone example illustrating the two approaches: https://gist.github.com/pgavlin/5eda82d8565b68b519a3aa6a3d20e86e

And here's the resulting assembly as generated with `clang -O3 -S` on OS X (in AT&T syntax, so destination registers come second): [get](https://gist.github.com/pgavlin/f8c36509fdfecc9647c4e3cee4c7fcb6#file-pointer-to-data-member-s-L5-L27) [get2](https://gist.github.com/pgavlin/f8c36509fdfecc9647c4e3cee4c7fcb6#file-pointer-to-data-member-s-L29-L56)

The option with a table of data members is one instruction shorter but requires a memory load; the option without the table is one instruction longer and contains two more branches but no memory accesses. Which of these is actually faster is a great question. I'll try asking a profiler :)